### ISSUE TITLE
Unable to add new product (server preview release)

### ISSUE THREAD

**Issue by teitrand** [patzly/grocy-android#642]
Cannot create new product from app. 

Works fine from grocy itself. 

Logs: (anonymized) 

--------- beginning of main
01-26 07:33:22.389 24785 26747 E BLASTBufferQueue: BLASTBufferItemConsumer::onDisconnect()
01-26 07:33:31.419 24785 26747 E BLASTBufferQueue: BLASTBufferItemConsumer::onDisconnect()
01-26 07:33:35.267 24785 26747 E BLASTBufferQueue: BLASTBufferItemConsumer::onDisconnect()
01-26 07:33:39.479 24785 24785 E SpannableStringBuilder: SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
01-26 07:33:39.479 24785 24785 E SpannableStringBuilder: SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
01-26 07:33:39.490 24785 24785 E SpannableStringBuilder: SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
01-26 07:33:39.490 24785 24785 E SpannableStringBuilder: SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
01-26 07:33:44.694 24785 26747 E BLASTBufferQueue: BLASTBufferItemConsumer::onDisconnect()
01-26 07:33:50.450 24785 26747 E BLASTBufferQueue: BLASTBufferItemConsumer::onDisconnect()
01-26 07:33:55.346 24785 27535 E Volley  : [151808] BasicNetwork.performRequest: Unexpected response code 400 for http://xxxxxxxxxxxxx/api/objects/products
01-26 07:33:55.365 24785 24785 E MasterProductViewModel: saveProduct: com.android.volley.ClientError
01-26 07:34:23.213 24785 26747 E BLASTBufferQueue: BLASTBufferItemConsumer::onDisconnect()
01-26 07:34:30.340 24785 26747 E BLASTBufferQueue: BLASTBufferItemConsumer::onDisconnect()
01-26 07:34:38.866 24785 26747 E BLASTBufferQueue: BLASTBufferItemConsumer::onDisconnect()
01-26 07:34:43.685 24785 26747 E BLASTBufferQueue: BLASTBufferItemConsumer::onDisconnect()
01-26 07:34:44.452 24785 26747 E BLASTBufferQueue: BLASTBufferItemConsumer::onDisconnect()
01-26 07:35:21.573 24785 26747 E BLASTBufferQueue: BLASTBufferItemConsumer::onDisconnect()
01-26 07:35:27.823 24785 26747 E BLASTBufferQueue: BLASTBufferItemConsumer::onDisconnect()

**Comment by patzly** [comment]
I cannot reproduce the issue, can you please provide more information? Like a screen recording of your actions or a precise description of what you do to reproduce the issue.

**Comment by teitrand** [comment]
I have tried to spin up a docker image and connect to it. Works fine.

The installation I currently use is a PHP install. Everything works fine besides creating any new product. I think the issue may be on my side of the table.

My PHP version is 8.2.



Sendt fra Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by teitrand** [comment]
I have tried to downgrade PHP version to 8.1. Still same result. Only adding products are being 
Tried the API for products with postman with no problems. Authorized and added product with statuscode 200. 

### Using GROCY ANDROID to add product:

**Steps to reproduce:** 

###  Grocy server
> Make sure grocy is freshly installed normally in a virtual server. 
> Delete database in "data" folder and reload grocy to reset all and generate new database.
> I use virtualmin. All requirements are met and site is running fine.  
> add api key for grocy server in menu
> No errors when using the grocy server itself

### Grocy Android
> Open grocy android
> Using QR-code to connect to server API.
> Go to Master data
> Click "products"
> Click plus-icon
> add name "Hobby nidar"
> add default location "Fridge"
> add quantity unit stock: "Piece"
> Notice the two bottom links are greyed out and says "Product is not on server".
> Click save icon

Error message "An error occurred"

**Apache Access Log:**
`91.186.71.2 - - [26/Jan/2023:19:36:07 +0100] "POST /api/objects/products HTTP/1.1" 400 437 "-" "Dalvik/2.1.0 (Linux; U; Android 12; DN2103 Build/SP1A.210812.016)"
`

**Product is not added to database and the error message will not let me continue.** 


### Using POSTMAN to add product:

URL for api is: `https://_<myprivateurl>_/api/objects/products`
Mode: `POST`

JSON: 
```
{
        "id": 76,
        "name": "Hobby Nidar",
        "description": "",
        "product_group_id": "",
        "active": 1,
        "location_id": 4,
        "shopping_location_id": "",
        "qu_id_purchase": 2,
        "qu_id_stock": 2,
        "min_stock_amount": 0,
        "default_best_before_days": 0,
        "default_best_before_days_after_open": 0,
        "default_best_before_days_after_freezing": 0,
        "default_best_before_days_after_thawing": 0,
        "picture_file_name": null,
        "enable_tare_weight_handling": 0,
        "tare_weight": 0,
        "not_check_stock_fulfillment_for_recipes": 0,
        "parent_product_id": null,
        "calories": 0,
        "cumulate_min_stock_amount_of_sub_products": 0,
        "due_type": 1,
        "quick_consume_amount": 1,
        "hide_on_stock_overview": 0,
        "default_stock_label_type": 0,
        "should_not_be_frozen": 0,
        "treat_opened_as_out_of_stock": 1,
        "no_own_stock": 0,
        "default_consume_location_id": "",
        "move_on_open": 0,
        "row_created_timestamp": "2023-01-26 08:31:42",
        "qu_id_consume": 2,
        "auto_reprint_stock_label": 0
    }
```

**Response for this is status code 200, and product is added to the database**


### OTHER info

Seems like all other POST actions are working; Location, Stores, Quantity units, product groups etc...  Name, quantity, location of product does not have anything to do with the failing POST request.

If I install to a docker container it works flawlessly. (Weird)

**Comment by dominiczedler** [comment]
As you are having the `auto_reprint_stock_label` key in your postman request, I think you use the unreleased preview version of grocy server. This will be grocy server 4.0.0 and like the major version jump indicates, it contains breaking changes in the API. The breaking key is `qu_factor_purchase_to_stock` which is missing then, and since our app sends the "unknown" key to the server this request is not valid for it.

I will fix it for the upcoming grocy version, but next time please give us the hint that you use a not supported future version. We can't always look for upcoming changes so most of the time we will fix breaking things after the server version was released.

**Comment by teitrand** [comment]
**I had to check, but my Grocy installation says this:**

```
Version 3.3.2
Utgitt 2022-11-12 2 måneder siden

PHP Version 8.1.14
SQLite Version 3.34.1
OS Linux 5.10.0-20-amd64 #1 SMP Debian 5.10.158-2 (2022-12-13) x86_64
Client Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36
```

**Comment by teitrand** [comment]
I am so sorry. It turns out that I had pulled the latest master, so you are right. I have been running a unreleased version. My bad. 
I downgraded and runned the composer- and yarn-script again and deleted the database. Now it it running perfectly again.

**Comment by dominiczedler** [comment]
No problem! 😄
You're always welcome.

I leave the issue open for me to support that breaking change in the API in some future release.

**Comment by dominiczedler** [comment]
Fixed with [a0536e2](https://github.com/patzly/grocy-android/commit/a0536e28db6cadc964310d2b0763ae5d315ce51f)
